### ISSUE TITLE
Explain weighted loss function for deterministic model

### ISSUE THREAD

**Issue by Joshua-Miller-161** [Joshua-Miller-161/mlde_Josh_zarr_NB#5]
@claude look at the repo and explain what the weighted loss function for the deterministic model does. Also look at scientific journals on the internet to see which paper it's based on

**Comment by Joshua-Miller-161** [comment]
@claude read this paper I uploaded. IN a similar fashion to the Hess and Boers loss function, add functionality for the Doury et al (2024) loss. It must be seemlessly integrated for use in the deterministic case just like the hess and boers loss I implemented is

[On the suitability of a convolutional neural network based RCM-emulator for fine spatio-temporal precipitation.pdf](https://github.com/user-attachments/files/25767563/On.the.suitability.of.a.convolutional.neural.network.based.RCM-emulator.for.fine.spatio-temporal.precipitation.pdf)

**Comment by Joshua-Miller-161** [comment]
@claude here's the text from the paper: 2.4 Loss function for the neural network training
Over this study, we propose a deeper look on the impact of
the loss function on the emulator’s performance. The loss
function is an essential part of the neural network training.
In the training phase, the network sees examples of inputs
and target pairs. For each day of the training set, it makes a
prediction and compares it with the truth. The loss function
evaluates the network prediction against the expected outcome.
The network parameters are then updated according
to the loss function results. This operation is repeated until
the cost (i.e. the loss mean on the training set) stabilises. The
best combination of parameters has the lowest cost over a
validation set, different from the training set. This is then a
minimisation problem to find the best estimate ̂F
such that:
Where  is the ensemble of possible parameters, V the validation
set and L the loss function.
(1) ̂F
= arg min
𝜃∈𝛩
L(V, 𝜃)
Precipitations are particularly complicated to emulate
with neural networks because of their distribution. Indeed,
as illustrated in Fig. 4, the distribution of precipitation
looks like a highly left-skewed gamma distribution. There
are many days with no precipitation and few ones with
very high precipitation, which induces heavy tail distributions.
These different events contribute non equally to the
mean, with a few days having more impact than the other
ones as illustrated on Fig. 4. It is of fundamental interest
that the emulator reproduces well the entire distribution.
The good reproduction of the frequency and intensity of
rare extreme events constitutes a substantial added value of
RCM, so the emulator should reproduce them accurately.
The loss function is therefore a possible way to rebalance
the data and to force the emulator to look more specifically
into some specific part of the distribution (Ayzel
et al. 2020).
We compare here three emulators, constructed with different
loss functions:
• Emul-MSE uses the classical mean squared error for
the loss function, as stated in Doury et al. (2023). It corresponds
to the L2 distance.
L(y, ŷ) = (2)
1
N × T
T
t=0

i∈D

yi,t − ̂yi,t
2
Fig. 4 Illustration of daily precipitation distribution (in mm/day), in
the Cevennes box (cf Fig. 2), all points and days are pooled. On the
left side we show the classical representation of the distribution while
the right side illustrate the distribution in term fractional contribution
to mean (explained in Sect. 2.5.3)
On the suitability of a convolutional neural network based RCM‑emulator for fine spatio‑temporal… 8593
With D the ensemble of grid points, N the number of
grid points and T the number of days.
• Emul-MAE uses the mean absolute error. It corresponds
to the L1 distance.
• Emul-ASYM uses a specific loss function designed for
the precipitation problem. It is based on the MAE loss
function plus an asymmetric term which penalizes the
emulator when it underestimates the true value while it
was a raining day. The stronger the rain the stronger the
penalty.
With i,t = Gi(yi,t) and Gi the cumulative distribution
function of a random variable Yi following a gamma
distribution
where the i and i parameters are fitted on the historical
precipitation series at each grid point i.
The MAE and MSE losses are the most commonly used loss
functions for regression problems. The MAE loss sums the
absolute distance between an observation and its prediction.
It gives the same weight to each observation. Knowing that
daily rainfalls are strongly left-skewed, with a vast number
of observations with a small amount of precipitation, the
EMUL-MAE should be able to fit these days well. However,
the rare cases with large precipitations could be less well
reproduced.
The MSE loss function gives more weight to the significant
errors than the small ones. The MSE generally shows
the best results in regression problems and is equivalent to
the maximum likelihood estimation in a Gaussian setting. It
leads theoretically to the best estimate for normally distributed
data knowing the inputs. In the case of precipitations,
it is not likely to be the case because of their highly intermittent
nature. So the MSE loss function might not be well
suited. Note that Emul-MSE is the same emulator as the one
introduced in Doury et al. (2023).
The choice of the asymmetric loss function comes from
the results of both EMUL-MAE and EMUL-MSE presented
in Sect. 3. The idea is to add a penalty when the emulator
underestimates strong precipitations. This is done by the
asymmetric term: max(0, yi,t − ̂yi,t) . Moreover it needs to
depend on the rain intensity. The more extreme the precipitation,
the rarest it is and so the higher the penalty should
L(y, ŷ) = (3)
1
N × T
T
t=0

i∈D
 
y
i,t
−
̂yi,t

(4)
L(y, ̂y) =
1
N × T
T
t=0

i∈D
 
y
i,t
−
̂yi,t

+ 𝛾2
i,t × max

0, yi,t − ̂yi,t

Yi ∼ i ∶ 

i, i

be. The i,t parameter determines how extreme is a given
observation and defines the weight accordingly. At each grid
point, we estimated the parameters of a gamma distribution
on the rainy days (over 1 mm) of the training set (using the
scipy python package, Virtanen et al. 2020). The Gamma
distribution has been widely used to described precipitation
data (Katz 1977; Vrac and Naveau 2007) but other distribution
could be considered. In order to make this parameter
estimation more robust, we fit them yearly and then average
these parameters over the years. It gives a map of the shape
and scale parameters. The i,t parameter is then the evaluation
of yi,t (the target value at point i and time t) by the
Cumulative Distribution Function (CDF) associated to the
gamma distribution i fitted for this point. It is an objective
way to indicate the relative intensity of the precipitation for a
given location. The asymmetric term acts like a “regularization”
parameter in the loss function as it aims to correct the
defaults of the MAE based emulator.
2.5 Evaluation metrics
In order to evaluate and compare the performances of the
emulators we will evaluate their predictions with respect to
the daily precipitation series from the corresponding RCM
simulation (cf Fig. 1). The evaluation relies on various metrics
to compare the targeted (Y) and the predicted ( ̂Y
) series
to have the most complete evaluation possible and understand
the strengths and weaknesses of each emulators. The
different metrics are detailed below.
2.5.1 Time series comparison
First of all we will evaluate in each grid point if the emulated
time series matches the original RCM series through
two metrics:
– Temporal Anomalies Correlation. This is the Pearson
correlation coefficient after removing the seasonal cycle:
with  the Pearson correlation coefficient and Ya and  Ya
are the anomaly series after removing a seasonal cycle
computed on the whole series.
– Ratio of Variance. It indicates the performance of the
emulator in reproducing the local daily variability. We
provide this score as a percentage:
Both metrics are computed at each grid point. Each map
is summarised with its spatial mean and 5th and 95th
ACC(Y, Y) =  (5)

Ya,  Ya

,
RoV(Y, Y) = (6)
Var

Y

Var(Y)
∗ 100
8594 A. Doury et al.
super-quantiles. The super-quantile  is defined as the mean
of all the values larger (resp. smaller) than the quantile of
order  , when  is larger (resp. smaller) than 0.5.
2.5.2 Climatological scale metrics
It is necessary to evaluate the emulators at the climatological
scale. We use three statistics over at least 20 years: the daily
precipitation mean, the 99th quantile and the percentage of
dry days (precipitations lower than 1 mm/day). These three
metrics, often used in the climate community, are snapshots
of the variable distribution illustrating the performances of
the emulators from low to high precipitation. The biases
maps are presented in percentage. When the biases are too
strong, notably because of comparing very small values, we
use the simple bias (̂Y
− Y) , expressed in mm/days. Again,
the statistics are computed point-wise, and each map is summarised
by its spatial mean and super-quantiles.
These three statistics will be looked at in present climate
but also in climate change context. Each statistic will be
computed in a future period and the climate change statistic
is the relative difference with the past period. Then the simple
bias is computed between RCM and emulator climate
change statistics.
2.5.3 PDF normalisation
Since the probability density function for daily precipitation
are very heavy-tailed, it is difficult to compare distribution
we get from the emulators with the RCM. We propose
here to use the ASoP method introduced in Klingaman
et al. (2017) and applied in multiple studies as Berthou
et al. (2020) or Vergara-Temprado et al. (2020). It consists
in computing the precipitation frequency following some
well-chosen bins bn defined in Eq. 8. The bins are such that
they contain a similar number of events for bins over 1 mm
and as long as the number of events is sufficient.
Then we can look at each bin’s contribution Cn to the mean
by multiplying each frequency by the corresponding bin’s
mean as described in Eq. 9. Both frequency and contribution
Relative bias = (7)

̂Y
− Y

Y
× 100
(8)
bn = e

log(0.005) +

n(log(120) − log(0.005))2
59
12

with n  [[0, 100]]
are interesting in comparing the emulated series with the
true RCM.
We use the skill score proposed in Berthou et al. (2020) to
evaluate the difference between the emulators and the RCM
truth contributions curves. The fractional contributions are
the actual contributions divided by the total mean precipitation
of the series. They give information on the shape of the
distribution independently from the mean. The Fractional
Contribution Skill Score (FCSS) sums the absolute difference
in each bin between the fractional contributions of an
emulator and the targeted true series. The area under the FC
curve is equal to 1, so the FCSS is equal to 0 when the two
distributions are identical and to 2 when there is no overlap
between them. It measures the differences between the two
distribution shapes independently from the series mean. This
score is illustrated on Fig. 9 and further commented in the
results Sect. 3.1.3.
2.5.4 SAL score
In order to further evaluate the performances of the emulator,
we use an object-oriented score introduced in Wernli
et al. (2008). The SAL score aims to evaluate the spatial
structure of precipitation objects from a predicted map versus
a reference. It compares two maps of precipitation at a
given time step. It accounts for the objects’ structure (S-component),
location (L-component) and the total amplitude of
precipitation (A-component). In perfect model evaluation,
the emulator should be able to reproduce the precipitation
events accurately. This score indicates if the emulator recreates
objects with the same characteristics than the RCM.
Note that the days are dealt independently meaning that the
life time of the objects is not considered.
The first step is to identify the precipitation objects. To
do so, we used the pysteps (Pulkkinen et al. 2019) python
library, which integrates a SAL implementation. On each
daily map, the objects are define as the groups of at least 5
consecutive points with precipitation higher than a threshold
equal to R∗ = 1
15
R(95) , R(95) being the 95th quantile on
the map. Multiple objects can be detected every day. Then,
the three components are computed aiming to differentiate
objectively different precipitation objects. The A- and
S- components take values between – 2 and 2 while the
(9)
Cn = fnmn
where fn and mn are the frequency and the mean of bin bn
(10)
FCSS(Emul, RCM) =

n∈[[0,100]]

FCEmul
n − FCRCM
n

where FCn =
Cn Σ
n Cn
=
Cn
mean
On the suitability of a convolutional neural network based RCM‑emulator for fine spatio‑temporal… 8595
L-component takes values between 0 and 2. If all objects
are similar on the maps the three components will be close
to 0. A more detailed presentation of the score behavior can
be find in Wernli et al. (2008, 2009).
The results are then presented in a diagram where each
day is represented by a point with the S and A components
on the x and y axis respectively, and the L component given
by the color of the point. SAL diagram are visible in Fig. 11
and commented in Sect. 3.1.4. Following the recommendation
of Wernli et al. (2009), we apply this score on subdomains
of a maximum of 500 km by side represented with
blue squares on Fig. 2. It is noteworthy that other metrics
could be applied to evaluate an RCM-emulator in an object
perspective through tracking (Caillaud et al. 2021) or spectral
density (Vosper et al. 2023) approaches.
3 Perfect model evaluation
This section is divided in two parts. In a first evaluation
step we evaluate and compare the three emulators in perfect
model framework. We use the CNRM-ALADIN RCP45
simulation, from 2006 to 2100, which has not been seen
during the training of the neural network (see Fig. 1). After
a first impression on the emulators’ abilities through some
examples, we extend the analysis with climatological and
daily scores. This section also aims to understand the impact
of the loss function on the trained emulator by comparing
the results of EMUL-ASYM with the two benchmarks emulators.
A second step focuses the evaluation on the Emul-
ASYM and the assessment of the transferability question by
extending the analysis to all available ALADIN63 simulations
(cf. Table 1).

**Comment by Joshua-Miller-161** [comment]
@claude The amount of rainfall data I have is ~30G. Make sure the fit_gamma_params function uses xarray or dask or something (but only optimize computations for 1CPU) to fit the rainfall distribution. Also include plenty of progress bars

**Comment by Joshua-Miller-161** [comment]
@claude This is looking better. Three changes. 1. The path were the parameters are stored must be the following: /WORK_DIR/doury_gamma_params/dataset_name/train_filename/parameters.extension    For the train_filename omit the extension.     WORK_DIR is from a .env file     2. This will be likely be running on only 1 CPU and the precipitation data is dask-backed with chunks <xarray.DataArray 'precipitation' (time: 213264, lat: 192, lon: 192)> Size: 31GB
dask.array<open_dataset-precipitation, shape=(213264, 192, 192), dtype=float32, chunksize=(512, 192, 192), chunktype=numpy.ndarray>
Coordinates:
  * time     (time) datetime64[ns] 2MB 2000-06-01 ... 2024-09-30T23:00:00
  * lat      (lat) float32 768B 50.95 50.85 50.75 50.65 ... 32.05 31.95 31.85
  * lon      (lon) float32 768B 2.55 2.65 2.75 2.85 ... 21.35 21.45 21.55 21.65
Attributes:
    units:    mm/hr                             use dask and/or xarray to parallelize for speed. Perhaps utilize LocalCluster for speedup and rechunk the data if necessary. If rechunked for gamma fitting, YOU MUST PUT BACK INTO ORIGINAL CHUNKS. ORIGINAL CHUNKING IS OPTIMIZED FOR MACHINE LEARNING.       3. Add a check so that before training the code will see if the correct parameters file for the corresponding training file exists and skips fitting if it does

**Comment by Joshua-Miller-161** [comment]
@claude if training.asym_train_filename = ''  # filename of training zarr, e.g. 'train_summer.zarr' in the config is left empty make it default to the FLAGS.filename in main.py     Usually I will NOT pass a separate file for fitting the gamma distribution and I want it to come from the training file I use for the machine learning
